internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/kafka/consumer/ApacheKafkaConsumerAdapter.java

```diff
@@ -42,100 +43,113 @@
+   * @throws IllegalArgumentException if the topic-partition is null or the partition number is negative
+   * @throws PubSubTopicDoesNotExistException if the topic does not exist
+   */
   @Override
   public void subscribe(PubSubTopicPartition pubSubTopicPartition, long lastReadOffset) {
     String topic = pubSubTopicPartition.getPubSubTopic().getName();
     int partition = pubSubTopicPartition.getPartitionNumber();
     TopicPartition topicPartition = new TopicPartition(topic, partition);
     Set<TopicPartition> topicPartitionSet = kafkaConsumer.assignment();
-    if (!topicPartitionSet.contains(topicPartition)) {
-      List<TopicPartition> topicPartitionList = new ArrayList<>(topicPartitionSet);
-      topicPartitionList.add(topicPartition);
-      kafkaConsumer.assign(topicPartitionList);
-      seekNextOffset(topicPartition, lastReadOffset);
-      assignments.put(topicPartition, pubSubTopicPartition);
-      LOGGER.info("Subscribed to Topic: {} Partition: {} Offset: {}", topic, partition, lastReadOffset);
+    if (topicPartitionSet.contains(topicPartition)) {
+      LOGGER.warn("Already subscribed on topic-partition:{}, ignoring subscription request", pubSubTopicPartition);
+      return;
+    }
+
+    // Check if the topic-partition exists
+    if (!isValidTopicPartition(pubSubTopicPartition)) {
+      LOGGER.error("Cannot subscribe to topic-partition: {} because it does not exist", pubSubTopicPartition);
+      throw new PubSubTopicDoesNotExistException(pubSubTopicPartition.getPubSubTopic());
+    }
+
+    List<TopicPartition> topicPartitionList = new ArrayList<>(topicPartitionSet);
+    topicPartitionList.add(topicPartition);
+    kafkaConsumer.assign(topicPartitionList); // add the topic-partition to the subscription
+    // Use the last read offset to seek to the next offset to read.
+    long consumptionStartOffset = lastReadOffset == OffsetRecord.LOWEST_OFFSET ? 0 : lastReadOffset + 1;
+    if (lastReadOffset == OffsetRecord.LOWEST_OFFSET) {
+      kafkaConsumer.seekToBeginning(Collections.singletonList(topicPartition));
     } else {
-      LOGGER
-          .warn("Already subscribed on Topic: {} Partition: {}, ignore the request of subscription.", topic, partition);
+      kafkaConsumer.seek(topicPartition, consumptionStartOffset);
+    }
+    assignments.put(topicPartition, pubSubTopicPartition);
+    LOGGER.info(
+        "Subscribed to topic-partition: {} with consumptionStartOffset: {}",
+        pubSubTopicPartition,
+        lastReadOffset);
+  }
+
+  private boolean isValidTopicPartition(PubSubTopicPartition pubSubTopicPartition) {
+    if (pubSubTopicPartition == null) {
+      throw new IllegalArgumentException("PubSubTopicPartition cannot be null");
+    }
+    if (pubSubTopicPartition.getPartitionNumber() < 0) {
+      throw new IllegalArgumentException("Partition number cannot be negative");
+    }
+
+    List<PubSubTopicPartitionInfo> topicPartitionInfos;
+
+    int retries = 5;
```

Comment:
>If it is second case (topic partition does not exist), checking them 5 times with 5 seconds probably will not change the behavior.

This is to account for intermittent issues with TP check. We don't want to block for too long either as this is shared consumer and could impact other subscription requests.

Reply:
1 second isn't short. We do multiple retries within 1 seconds. And logging will pollute without much help

Reply:
Essentially, I just do not want to hide any potential bug here. For those retries, I just worried it will block other partition consuming. During server startup, we have a lot of partitions to be subscribed for each shared consumer, I just do not know blocking serveral seconds for every not-existent topic partition is ideal or not.

Reply:
> Essentially, I just do not want to hide any potential bug here. 

What kind of bugs?

Reply:
Discussed offline and resolved.